packages/web/src/application/test/zoom/zoom-provider.test.tsx

```diff
@@ -0,0 +1,96 @@
+import React, { ReactNode } from 'react';
+import { MemoryRouter } from 'react-router-dom';
+import { render, screen, waitFor } from '@testing-library/react';
+import { User } from '@chess-tent/models';
+import application from '@application';
+import { users, zoom } from '../fixtures';
+
+const { components } = application;
+
+interface RenderOptions {
+  children: ReactNode;
+  user: User;
+  meetingNumber?: string;
+  queryCode?: boolean;
+}
+
+const renderInsideProvider = ({
+  children,
+  user,
+  meetingNumber,
+  queryCode,
+}: RenderOptions) => {
+  const { ZoomProvider } = components;
+
+  render(
+    <MemoryRouter
+      initialEntries={[
+        { pathname: '/', search: queryCode ? '?code=testcode' : undefined },
+      ]}
+    >
+      <ZoomProvider
+        redirectUri={zoom.redirectUri}
+        user={user}
+        meetingNumber={meetingNumber}
+      >
+        {children}
+      </ZoomProvider>
+    </MemoryRouter>,
+  );
+};
+
+describe('Zoom Provider', () => {
+  const coach: User = users.find(user => user.coach)!;
+  const student: User = users.find(user => !user.coach)!;
+
+  test('Student should have only password input', async () => {
+    const { ZoomGuestControl } = components;
+
+    renderInsideProvider({
+      children: <ZoomGuestControl />,
+      user: student,
+      meetingNumber: zoom.meetingNumber,
+    });
+
+    const passwordInput = screen.queryByPlaceholderText(
+      'Meeting password (if any)',
+    );
+    const meetingNumberInput = screen.queryByPlaceholderText('Meeting number');
+
+    await waitFor(() => expect(passwordInput).toBeTruthy());
+    await waitFor(() => expect(meetingNumberInput).toBeFalsy());
+  });
+
+  test('Coach should have authorize button only when no authCode', async () => {
+    const { ZoomHostControl } = components;
+
+    renderInsideProvider({
+      children: <ZoomHostControl />,
+      user: coach,
+    });
+
+    const authButton = screen.getByRole('button', {
+      name: 'Authorize Zoom',
+    });
+
+    await waitFor(() => expect(authButton).toBeTruthy());
```

Comment:
This is good and in addition you want to test logic/integration.

**This is important catch when testing, read it carefully!**

Testing logic means you're testing certain behaviour that component provides, that's added value of the component/function itself.

Testing integration means you're using a function/component that does the logic and you want to test if it's integrated in the correct way. You don't test the function behaviour because you assume it's already tested.

This is also a difference between black and white box tests.

If you're using `authoriseZoom`, and it's already tested, you can in this case only test if the `authoriseZoom` has been called with the correct arguments.

One other approach, this applies to smaller pieces that only exists or are used only inside component, not anywhere else, so they are extracted to functions only to make code more readable, is to no test those functions independently but rather through the component/parent.
This way you keep black box testing, you're testing the behaviour and not the integration but the function itself doesn't have unit tests.

Reply:
Should we split the tests and name folders accordingly, or is it ok like this?

Reply:
It's ok like this.